mDNSWindows/mDNSWin32.c

```diff
@@ -1054,7 +1054,7 @@ mDNSPlatformTCPConnect
 
 	// Setup connection data object
 
-	sock->userCallback	= inCallback;
+	sock->userCallback	= (TCPUserCallback) inCallback;
```

Comment:
Your response to my question about this was to say that it should be cast to the right type pointer, but that's not right. The callback should be of the right type. There shouldn't be a cast here at all. This is not safe coding.

Reply:
Not precisely a "right" pointer type, my point was primarily saying this is a wacky way to simulate lambdas in C. And it was just my guess of intentions of whoever wrote this code.

Having said that, I'm happy to change it to the correct pointer type. Indeed after some more in-depth code searching it looks like this specific callback isn't used the way I assumed it is (i.e. like simulated lambdas) and it should work to just replace TCPUserCallback with TCPConnectionCallback. I'll jump on it as soon as you give me green light on all other changes here :). I would really like to see at least some of them merged.

Reply:
Sounds good. I too would like to get this stuff merged—sorry it's taken so long!

Reply:
Am I waiting on you for this change, or did I drop the ball somewhere? :)

Reply:
This time it's on me - I misunderstood one of your previous comments, sorry!

I tried making the change we discussed, but it turned out the callback is indeed used as a quirky C-style lambda in two places with two different argument sets: `mDNSWindows/SystemService/Service.c` and `mDNSWindows/mDNSWin32.c`. We could use an union here to avoid casting (PTAL how I did it on top of this change: https://pastebin.com/Dgyunh9F). Please let me know if that's what you're looking for so I would apply the linked diff on top of this change.

I know it would be best to use discriminated unions (i.e. `std::variant`) or just lambdas/bind expressions (i.e. `std::function`), but this project is C-only, I'm afraid :).

Reply:
Sorry for my confused comment on an earlier discussion. I ran into this overloading when I updated mDNSPosix to support TCP a few years ago. There is actually (as far as I can see) no reason to do this overloading. These are two different callbacks with two different function signatures. They aren't interchangeable. So the fix is easy: have two different callbacks in the TCPSocket structure.

In the Posix code (and the MacOSX code) uds_daemon uses a unix domain socket, not a TCP socket, so we don't see this conflict. You could do this in the Windows code as well—the fact that both of these structures are managing TCP connections doesn't mean there can only be one structure.

I would offer to just fix this myself, but I don't have a Windows machine to test on. Would you mind trying one of these solutions?

Reply:
Makes sense. I started with using separate callbacks in the first place, but then used union to avoid any potential ABI incompatibilities. So the final change will be very similar to https://pastebin.com/Dgyunh9F except the two callbacks will not be grouped into an union. Let me push the updated change shortly.

Reply:
Pushed, PTAL.

Reply:
Great, thanks! I've integrated the other pull requests in our repo, so this is the last one, and then I can sync to the github repo.

Reply:
Thanks, when you have a chance, PTAL at the spreadsheet I linked at https://github.com/Abhayakara/mdnsresponder/pull/14 for the potential next batch of patches. There's a few more or less obvious bugs and a couple of feature requests.